### ISSUE TITLE
Problem handling environment variables with colon

### ISSUE THREAD

**Issue by yookoala** [fish-shell/fish-shell#4039]
Version: Fish 2.3.1

Here are 2 form of what I believe the same problem

# Problem 1: Universal Variables
## Steps to Reproduce
1. Start a fish session and type:
```fish
set -Ux SOME_PATH "/home/foobar/foo:/home/foobar/bar"
```
2. Close that fish session. Then start another one and type:
```fish
echo $SOME_PATH
```

## Expected
See something like this:
```
/home/foobar/foo:/home/foobar/bar
```

## See Instead
```
/home/foobar/foo /home/foobar/bar
```
Note: the space between the 2 paths is a special character. Not really a space.

# Problem 2: Fish script in conf.d
## Steps to Reproduce
1. Open `$HOME/.config/fish/conf.d/foobar.fish and 
```fish
#!/usr/bin/fish
set -gx SOME_PATH "/home/foobar/foo:/home/foobar/bar"
```
2. Close that fish session. Then start another one and type:
```fish
echo $SOME_PATH
```

## Expected
See something like this:
```
/home/foobar/foo:/home/foobar/bar
```

## See Instead
```
/home/foobar/foo /home/foobar/bar
```
Note: the space between the 2 paths is a special character. Not really a space.

**Comment by faho** [comment]
Did you actually try it with "SOME_PATH", or did you try it with "PATH"?

Are you actually kust running 2.3.1, or is there some older fish version involved as well?

For me, this actually works. If I create a variable (even a universal one) called "SOME_PATH" (or really anything except for "CDPATH", "MANPATH" or "PATH"), it can include colons and won't be split.

What you are seeing here sounds like the behavior of fish before 2.2.0, where we would split inherited variables on colons, so a variable called "SOME_PATH" with a value of "foo:bar" would be turned into a list "SOME_PATH" with two elements "foo" and "bar". This would possibly also happen if you had these older fish versions running at the same time as newer ones.

To confirm that it is actually a list, use `printf '%s\n' $SOME_PATH` - this will insert a newline between the arguments, so if $SOME_PATH expands to two arguments, they'll be written on two lines.

**Comment by krader1961** [comment]
Like @faho I can't reproduce your problem. And I tested with fish 2.3.1, 2.5.0 and git head. Like @faho I suspect where you wrote `SOME_PATH` you're actually using `PATH` in which case the splitting is expected.

**Comment by yookoala** [comment]
I am actually having problem with `GOPATH` in shell. Not `PATH`.

Guess that can be issue from `rvm` or `nvm` integration. I am doing some more testing. Will update this post soom. Thanks for helping.

**Comment by krader1961** [comment]
I don't know why `rvm` or `nvm` would alter the `GOPATH` env var and can't recall anyone reporting this before. But there is definitely something custom in your fish config that is doing that transformation. Because nothing in the standard fish code does so.

Hold on! You wrote

> Note: the space between the 2 paths is a special character. Not really a space.

If you do `echo $GOPATH | xxd` or `echo $GOPATH | od -tx1z` what do you see? Is there a \x20 between the two paths? Because that is the character used by fish internally to separate array elements. It should never be present when a var is expanded in a command. Also, what does `count $GOPATH` report?

**Comment by yookoala** [comment]
Here are the results:

```
$ echo $GOPATH | xxd
00000000: 2f68 6f6d 652f 666f 6f62 6172 2f66 6f6f  /home/foobar/foo
00000010: 202f 686f 6d65 2f66 6f6f 6261 722f 6261   /home/foobar/ba
00000020: 720a
```

```
$ echo $GOPATH | od -tx1z
0000000 2f 68 6f 6d 65 2f 66 6f 6f 62 61 72 2f 66 6f 6f  >/home/foobar/foo<
0000020 20 2f 68 6f 6d 65 2f 66 6f 6f 62 61 72 2f 62 61  > /home/foobar/ba<
0000040 72 0a                                            >r.<
0000042
```

**Comment by krader1961** [comment]
Argh! What was I thinking. A \x20 is a simple space character. Fish uses \x1e as its internal array element separator. So the expansion of `$GOPATH` is what I would expect if the var has two elements (i.e., `count $GOPATH` outputs 2). Can you clarify what you meant when you wrote the following:

> Note: the space between the 2 paths is a special character. Not really a space.

**Comment by yookoala** [comment]
Go will look for libraries from `$GOPATH` paths. Also `GOPATH` supports multiple path. All you need to do is use colon to separate the paths.

`go` supports a command called `go get` for library installation. It will automatically grab code from github or other git source into `$GOPATH/src`. If there are multiple paths, the first path will be used. So if `$GOPATH` is "`/home/foobar/foo:/home/foobar/bar`", libraries will be downloaded into "`/home/foobar/foo/src`"

The odd thing is when I do `go get`, the library flows into `/home/foobar/foo\x1e/home/foobar/bar/src`. I have no idea why. Alternatively, it can be go's toolchians issue. I remember having no problem with go < 1.7.

**Comment by krader1961** [comment]
Thanks for the information but what the Go tools do with the var isn't important. What is important is that `go get` is getting the internal representation of the fish array. What that means is that you are, in fact, doing the equivalent of

```
set GOPATH /home/foobar/foo /home/foobar/bar
```

Notice the space rather than a colon. That creates a fish array with two elements. It is exported with a \x1E between the element so that fish subshells will properly parse the var. But other programs, like `go get` will not do the right thing since they expect a colon rather than a \x1E character to separate the path elements.

In other words, it seems unlikely you are doing `set -Ux GOPATH "/home/foobar/foo:/home/foobar/bar"`. Because that works as expected.

Perhaps you have a function defined with `--on-variable GOPATH` (or `-v GOPATH`) that is splitting the var on colons into a multi-element array.

**Comment by yookoala** [comment]
Indeed, I tried this:
```
$ set -xU GOPATH /home/foobar/foo:/home/foobar/bar
$ echo $GOPATH
/home/foobar/foo:/home/foobar/bar
```

Things go wrong only if I restart the terminal, then:
```
$ echo $GOPATH
/home/foobar/foo /home/foobar/bar
```

**Comment by krader1961** [comment]
There is definitely something in your *~/.config/fish/config.fish* or a script that it runs which is splitting the var into a multi element array. Try `cd ~/.config/fish` followed by `grep GOPATH **`. I'm betting you'll find some fish script that is rewriting the var.

**Comment by yookoala** [comment]
Turns out it seems the [rvm-fish integration](https://rvm.io/integration/fish) messes with my paths. If I disable it, the `GOPATH` went back to normal.

So this is not an issue for Fish. I'll close this issue now.

**Comment by yookoala** [comment]
@krader1961 Thanks for helping :-)

**Comment by krader1961** [comment]
Happy to help. The fact that `rvm` is mutating vars unrelated to Ruby is a serious bug. Definitely complain to whomever wrote that tool.

**Comment by krader1961** [comment]
Aha! These are the lines line [rvm.fish](https://raw.githubusercontent.com/lunks/fish-nuggets/master/functions/rvm.fish) which are causing your problem:

```
  # apply rvm_* and *PATH variables from the captured environment
  and eval (grep -E '^rvm|^[^=]*PATH|^GEM_HOME' $env_file | grep -v '_clr=' | sed '/^[^=]*PATH/s/:/" "/g; s/^/set -xg /; s/=/ "/; s/$/" ;/; s/(//; s/)//')
```

Notice that it is matching any var with a `PATH` suffix. That is not just wrong -- it is dangerous. As you found out. If the author of that fish script doesn't respond quickly to you bug report I would recommend finding a similar function written by someone more competent.
